x509-cert/src/name.rs

```diff
@@ -2,17 +2,181 @@
 
 use crate::attr::AttributeTypeAndValue;
 use alloc::vec::Vec;
+use const_oid::{
+    db::{rfc3280, rfc4519},
+    ObjectIdentifier,
+};
 use core::{fmt, str::FromStr};
-use der::{asn1::SetOfVec, Encode};
+use der::{
+    asn1::{Any, PrintableStringRef, SetOfVec},
+    Encode,
+};
 
 /// X.501 Name as defined in [RFC 5280 Section 4.1.2.4]. X.501 Name is used to represent distinguished names.
 ///
 /// ```text
 /// Name ::= CHOICE { rdnSequence  RDNSequence }
 /// ```
 ///
+/// # Example
+///
+/// ```
+/// use std::str::FromStr;
+/// use x509_cert::name::Name;
+///
+/// let subject = Name::from_str("CN=example.com").expect("correctly formatted subject");
+/// ```
+///
 /// [RFC 5280 Section 4.1.2.4]: https://datatracker.ietf.org/doc/html/rfc5280#section-4.1.2.4
-pub type Name = RdnSequence;
+#[cfg_attr(feature = "arbitrary", derive(arbitrary::Arbitrary))]
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct Name(RdnSequence);
+
+impl_newtype!(Name, RdnSequence);
+
+impl Name {
+    /// Is this [`Name`] empty?
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    /// Returns the number of [`RelativeDistinguishedName`]
+    pub fn len(&self) -> usize {
+        self.0 .0.len()
+    }
+
+    /// Iterator over the inner [`AttributeTypeAndValue`]
+    #[inline]
+    pub fn iter(&self) -> impl Iterator<Item = &'_ AttributeTypeAndValue> + '_ {
+        self.0 .0.iter().flat_map(move |rdn| rdn.0.as_slice())
+    }
+
+    /// Iterator over the inner [`RelativeDistinguishedName`]
+    #[inline]
+    pub fn iter_rdn(&self) -> impl Iterator<Item = &'_ RelativeDistinguishedName> + '_ {
+        self.0 .0.iter()
+    }
+}
+
+impl Name {
+    /// Returns the element found in the name identified by `oid`
```

Comment:
```suggestion
    /// Returns the element found in the name identified by `oid`.
```

Also, what happens if a `Name` contains multiple elements with the same OID?
- If that is allowed in the relevant RFC, then this method (and all methods calling it internally) need to explicitly say that they only return the first such element found, and later elements should be fetched via [`Self::iter`].
- If that is not allowed in the relevant RFC, then we should ensure that `Name` parsing does not allow it by default (and then these methods are fine as-is), maybe with a hazmat option reachable via the specific choice of certificate's profile.